### PR TITLE
Pangu Improvements

### PR DESCRIPTION
**Pull request by dallasfoster**
# Modulus Pull Request

## Description
This PR adds the following features/changes to the Pangu model and training script:

1. Configurable number of constant, surface, and atmosphere variables in the model.
2. Configurable number of upsampled and downsampled transformer blocks.
3. Gradient checkpointing support in the Pangu processor (encoder/decoder) layers.
4. Improved training script with improved static capture support, multistep rollout, validation function, and weighted loss function.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

**Comment by dallasfoster** [comment]
Depends on #660

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,10 +12,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Graph Transformer processor for GraphCast/GenCast.
 - Utility to generate STL from Signed Distance Field.
+- Improved Pangu training code
 
 ### Changed
 
 - Refactored CorrDiff training recipe for improved usability
+- Refactored Pangu model for better extensibility and gradient checkpointing support.
+    Some of these changes are not backward compatible.
```
Comment (pzharrington) [inline CHANGELOG.md:21]: Perhaps comment on what specifically is not backward compatible? Is it just the removal of the `prepare_input` routine from the Pangu model?

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 2]
```diff
@@ -203,122 +262,197 @@ def main(cfg: DictConfig) -> None:
...
+            invar_ = invar[b : b + 1]
+            cos_zenith_ = cos_zenith[b : b + 1]
+            for t in range(outvar.shape[1]):
+                invar_ = torch.concat(
+                    [surface_mask, cos_zenith_[:, t : t + 1], invar_], dim=1
+                )
+                outpred = my_model(invar_)
+                loss += loss_func(outpred, outvar[b : b + 1, t], weights) / batch_size
+                invar_ = outpred
+
```
Comment (pzharrington) [inline examples/weather/pangu_weather/train_pangu_era5.py:330]: Is there a reason this inner multistep loop cannot be run with batch_size > 1? Would be nice to support batched rollout training if it could fit in memory

File: examples/weather/pangu_weather/train_pangu_era5.py [hunk 3]
```diff
@@ -203,122 +262,197 @@ def main(cfg: DictConfig) -> None:
             )
         torch.cuda.current_stream().wait_stream(ddps)
 
+    # pangu_model = torch.compile(pangu_model, mode = "max-autotune")
```
Comment (pzharrington) [inline examples/weather/pangu_weather/train_pangu_era5.py:265]: Can drop these commented out `.compile` statements if they're unused

File: modulus/models/pangu/pangu_processor.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+
+import torch
+
+from ..layers import DownSample3D, FuserLayer, UpSample3D
+from ..module import Module
```
Comment (pzharrington) [inline modulus/models/pangu/pangu_processor.py:22]: Prefer direct imports here

**Comment by pzharrington** [comment]
I added some minor suggestions, overall this looks great though. I found the LambdaLR scheme with the custom hydra resolver a bit convoluted, maybe a ConstantLR would be simpler and more readable to achieve the same effect. However it is nice to have the example in there if someone wants to do more custom scheduling.
